src/main/java/com/google/googleidentity/oauth2/validator/AuthorizationEndpointRequestValidator.java

```diff
@@ -0,0 +1,145 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.validator;
+
+import com.google.common.base.Strings;
+import com.google.googleidentity.oauth2.client.ClientDetails;
+import com.google.googleidentity.oauth2.client.ClientDetailsService;
+import com.google.googleidentity.oauth2.exception.InvalidRequestException;
+import com.google.googleidentity.oauth2.exception.OAuth2Exception;
+import com.google.googleidentity.oauth2.exception.InvalidScopeException;
+import com.google.googleidentity.oauth2.exception.AccessDeniedException;
+import com.google.googleidentity.oauth2.exception.UnauthorizedClientException;
+import com.google.googleidentity.oauth2.exception.UnsupportedResponseTypeException;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.googleidentity.oauth2.util.OAuth2Utils;
+
+
+import javax.servlet.http.HttpServletRequest;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.Optional;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+public class AuthorizationEndpointRequestValidator {
+
+    private static final Logger log = Logger.getLogger("AuthorizationEndpointParameterValidator");
+
+
+    /**
+     * Check whether the uri is valid in a Get request to authorization endpoint
+     */
+    public static void validateRedirectUri(
+            HttpServletRequest request,
+            ClientDetailsService clientDetailsService) throws InvalidRequestException {
+        String clientID = request.getParameter(OAuth2ParameterNames.CLIENT_ID);
+
+        if (Strings.isNullOrEmpty(clientID)) {
+            throw new InvalidRequestException(InvalidRequestException.ErrorCode.NO_CLIENT_ID);
+        }
+
+        Optional<ClientDetails> client = clientDetailsService.getClientByID(clientID);
+
+        if (!client.isPresent()) {
+            throw new InvalidRequestException(
+                    InvalidRequestException.ErrorCode.NONEXISTENT_CLIENT_ID);
+        }
+
+        String redirectUri = request.getParameter(OAuth2ParameterNames.REDIRECT_URI);
+
+        if (Strings.isNullOrEmpty(redirectUri)) {
+            throw new InvalidRequestException(InvalidRequestException.ErrorCode.NO_REDIRECT_URI);
+        }
+
+        try {
+            redirectUri = URLDecoder.decode(redirectUri, "utf-8");
+        } catch (UnsupportedEncodingException e) {
+            log.log(Level.INFO, "Uri decode failed", e);
```

Comment:
nit: add a comment explaining that we proceed with the encoded redirectUri here rather than failing the request